internal/btf/ext_info.go

```diff
@@ -98,156 +96,268 @@ func parseBTFExtCoreHeader(r io.Reader, bo binary.ByteOrder, extHeader *btfExtHe
-		return extInfo{}, fmt.Errorf("ext_info record size mismatch, want %d (got %d)", ei.recordSize, other.recordSize)
-	}
+	for n, fi := range infos {
+		// Iterator went past the offset the caller is looking for,
+		// no point in continuing the search.
+		if offset < fi.InsnOff {
+			return nil
+		}
 
-	records := make([]extInfoRecord, 0, len(ei.records)+len(other.records))
-	records = append(records, ei.records...)
-	for _, info := range other.records {
-		records = append(records, extInfoRecord{
-			InsnOff: info.InsnOff + offset,
-			Opaque:  info.Opaque,
-		})
+		// If there is no next item in the list, or if the given offset
+		// is smaller than the next function, the offset belongs to
+		// the current function.
+		if n+1 >= len(infos) || offset < infos[n+1].InsnOff {
+			return &fi
+		}
 	}
-	return extInfo{ei.byteOrder, ei.recordSize, records}, nil
+
+	return nil
 }
 
-func (ei extInfo) MarshalBinary() ([]byte, error) {
-	if ei.byteOrder != internal.NativeEndian {
-		return nil, fmt.Errorf("%s is not the native byte order", ei.byteOrder)
+// parseLineInfos parses a func_info sub-section within .BTF.ext ito a map of
+// func infos indexed by section name.
+func parseFuncInfos(r io.Reader, bo binary.ByteOrder, strings stringTable) (map[string]FuncInfos, error) {
+	recordSize, err := parseExtInfoRecordSize(r, bo)
+	if err != nil {
+		return nil, err
 	}
 
-	if len(ei.records) == 0 {
-		return nil, nil
+	result := make(map[string]FuncInfos)
+	for {
+		secName, infoHeader, err := parseExtInfoSec(r, bo, strings)
+		if errors.Is(err, io.EOF) {
+			return result, nil
+		}
+
+		records, err := parseFuncInfoRecords(r, bo, recordSize, infoHeader.NumInfo)
+		if err != nil {
+			return nil, fmt.Errorf("section %v: %w", secName, err)
+		}
+
+		result[secName] = records
 	}
+}
+
+// parseFuncInfoRecords parses a stream of func_infos into a funcInfos.
+// These records appear after a btf_ext_info_sec header in the func_info
+// sub-section of .BTF.ext.
+func parseFuncInfoRecords(r io.Reader, bo binary.ByteOrder, recordSize uint32, recordNum uint32) (FuncInfos, error) {
+	var out FuncInfos
 
-	buf := bytes.NewBuffer(make([]byte, 0, int(ei.recordSize)*len(ei.records)))
-	for _, info := range ei.records {
-		// The kernel expects offsets in number of raw bpf instructions,
-		// while the ELF tracks it in bytes.
-		insnOff := uint32(info.InsnOff / asm.InstructionSize)
-		if err := binary.Write(buf, internal.NativeEndian, insnOff); err != nil {
-			return nil, fmt.Errorf("can't write instruction offset: %v", err)
+	var fi FuncInfo
+	for i := uint32(0); i < recordNum; i++ {
+		if err := binary.Read(r, bo, &fi); err != nil {
+			return nil, fmt.Errorf("can't read function info: %v", err)
 		}
 
-		buf.Write(info.Opaque)
+		if fi.InsnOff%asm.InstructionSize != 0 {
+			return nil, fmt.Errorf("offset %v is not aligned with instruction size", fi.InsnOff)
+		}
+
+		// Only the first 8 bytes of a funcInfo record are recognized,
```

Comment:
Good catch. This check also only needs to happen once, since the size of our func/lineinfo struct and `recordSize` don't change. Moved the check to the top of `parse{Func,Line}InfoRecords`.